src/PackagePrivate/MediaWikiNormalizedTermCleaner.php

```diff
@@ -0,0 +1,271 @@
+		// $this->dbr and $this->dbw are lazily initialized in cleanTerms()
+		$this->logger = $logger ?: new NullLogger();
+	}
+
+	/**
+	 * Delete the specified term_in_lang rows from the database,
+	 * as well as any text_in_lang and text rows that are now unused.
+	 *
+	 * @param int[] $termInLangIds
+	 */
+	public function cleanTerms( array $termInLangIds ) {
+		if ( $this->dbr === null ) {
+			$this->dbr = $this->lb->getConnection( ILoadBalancer::DB_REPLICA );
+			$this->dbw = $this->lb->getConnection( ILoadBalancer::DB_MASTER );
+		}
+
+		try {
+			$this->lb->beginMasterChanges( __METHOD__ );
+			$this->cleanTermInLangIds( $termInLangIds );
+			$this->lb->commitMasterChanges( __METHOD__ );
+		} catch ( DBError $exception ) {
+			$this->lb->rollbackMasterChanges( __METHOD__ );
+			$this->logger->error(
+				'{method}: DBError while cleaning terms {termInLangIds}: {exception}',
+				[
+					'method' => __METHOD__,
+					'termInLangIds' => $termInLangIds,
+					'exception' => $exception
+				]
+			);
+			throw $exception;
+		}
+	}
+
+	/**
+	 * Delete the specified term_in_lang rows from the database,
+	 * as well as any text_in_lang and text rows that are now unused.
+	 *
+	 * @param int[] $termInLangIds
+	 */
+	private function cleanTermInLangIds( array $termInLangIds ) {
+		if ( $termInLangIds === [] ) {
+			return;
+		}
+
+		$this->logger->debug(
+			'{method}: deleting {count} term_in_lang rows',
+			[
+				'method' => __METHOD__,
+				'count' => count( $termInLangIds ),
+			]
+		);
+
+		$potentiallyUnusedTextInLangIds = $this->selectFieldValuesForPrimaryKey(
+			'wbt_term_in_lang',
+			'wbtl_text_in_lang_id',
+			'wbtl_id',
+			$termInLangIds,
+			__METHOD__
+		);
+
+		$this->dbw->delete(
+			'wbt_term_in_lang',
+			[ 'wbtl_id' => $termInLangIds ],
+			__METHOD__
+		);
+
+		$stillUsedTextInLangIds = $this->dbw->selectFieldValues(
+			'wbt_term_in_lang',
+			'wbtl_text_in_lang_id',
+			[ 'wbtl_text_in_lang_id' => $potentiallyUnusedTextInLangIds ],
+			__METHOD__,
+			[
+				/**
+				 * If we try to clean up a text_in_lang whose last use in a term_in_lang we just
+				 * removed, and simultaneously another request adds a new term_in_lang using that
+				 * text_in_lang, we want one of the following to happen:
+				 *
+				 * 1. Our transaction completes first and removes the text_in_lang. The concurrent
+				 *    request blocks until we’re done, then sees that the text_in_lang is gone and
+				 *    creates it (again) as part of inserting the term_in_lang.
+				 * 2. The other transaction completes first and registers another term_in_lang using
+				 *    that text_in_lang. We block until that’s done and then notice the text_in_lang
+				 *    is still used and don’t clean it up.
+				 *
+				 * For this to work, we need to use 'FOR UPDATE' when checking whether a
+				 * text_in_lang is still used in a term_in_lang, and the other request needs to
+				 * check during or after insert of the new term_in_lang if the text_in_lang still
```

Comment:
With “during or after insert”, I’m currently thinking of two options:

- Do it directly in the query, with something like `… WHERE EXISTS(SELECT 1 FROM wbt_text_in_lang WHERE wbxl_id = $ID)`; if this did not insert anything, the `text_in_lang` was deleted, and we should acquire a new ID and repeat the insert with that.
- Insert unconditionally, then do a `SELECT 1 FROM wbt_text_in_lang WHERE wbxl_id = $ID` afterwards, and if the ID was deleted, insert it again, using the same ID so that we don’t need to touch `wbt_term_in_lang` a second time.

But we can probably think more about that later.

Reply:
I would replace "check" with "ensure"… the other process could also use `FOR UPDATE` (which I think is you first option)… no strict need to manually check (although that might also be an option).

Reply:
My first option was to do this on the `INSERT`, so I don’t think `FOR UPDATE` could be used there. I think what you’re proposing is a third option that I didn’t think of – I’m not sure what you want to do the `FOR UPDATE` on, but I’ll update the comment, “ensure” instead of “check” sounds good anyways.

Reply:
(The whole comment should be revisited anyways once we actually implement the other half of the race condition handing.)